v2.2/transaction-retry-errors.md

```diff
@@ -0,0 +1,152 @@
+independent actor that doesn’t need to concern itself with the
+internal behavior of any other transaction — only their final effect
+on the database.
+
+For this reason, many users are confused when some databases, like
+CockroachDB but also PostgreSQL, tell them that they need to be
+prepared to re-issue their query if something goes wrong.
+
+This behavior is counter-intuitive to the notion that the database
+could handle isolation as a service to the client. Does a retry error
+indicate the database is insufficiently or incorrectly able to provide
+isolation? Isn't it the database's job to, like, not do that? Or at
+least be able to sort it out itself if there is a doubt?
+
+There are two distinct questions here:
+
+- why can’t the db solve this problematic state own its own after it occurs, and
+- why can’t the db prevent me from getting into this problematic state in the first place?
+
+# Atomicity with interactive transactions
+
+The hurdle causing (1) is really that the database doesn’t control the
+rest of the universe. This is more about maintaining atomicity
+(reminder: the **A** in ACID) than it is about “having another go at
+it”. When the database decides the transaction can no longer be
+committed for some reason, it needs to make sure that everything that
+happened as a result of that transaction is undone.
+
+This includes reverting decisions made client-side using the data
+**returned by the database while the txn was open**, such as:
+
+- whether to advance the program counter in the driving application,
+- whether to send an email to a user,
+- whether to send a package through the postal service.
+
+In a traditional conversational SQL model, there’s no way for the
+database to guarantee that these things will be rolled back without
+informing the user that they need to do it themselves. In most cases
+this means requiring the application to not take irreversible actions
+until the transaction has successfully committed; but there is no way
+to enforce this in the SQL protocol.
+
+Some other SQL engines will avoid this problem by **preventing
+interactive transactions**, i.e. disallow a client from observing the
+result of an intermediate statement before it issues another statement
+while the txn is open. When a SQL engine only supports non-interactive
+transactions, it can freely decide to delay reporting the results
+until a txn is successfully committed, and handle txn retries
+internally itself. CockroachDB does not do this because it supports
+the PostgreSQL dialect which mandates support for interactive
+transactions.
+
+# Transaction scheduling
+
+What is causing hurdle (2) above?
+
+Shouldn't the database have a sophisticated transaction scheduler?
+
+If it had, it would intelligently provide concurrency, make
+transactions run as soon as theoretically possible, but prevent wrong
+orderings not allowed by serializability.
+
+The difficulty with this is twofold:
+
+1. To decide a good ordering requires a precise
```

Comment:
In my opinion, this section is a bit too in-depth and also it doesn't mention a bigger deal than the difficulties of dependency analysis - the fact that transactions are dynamic, as the model is conversational. So you don't get all the dependencies from the beginning. At which point, I'm not sure having a separate section from the previous one makes sense.